netlogo-gui/src/main/agent/ArrayAgentSet.scala

```diff
@@ -0,0 +1,245 @@
+      var i = 0
+      while (i < random) {
+        iter.next()
+        i += 1
+      }
+      iter.next()
+    }
+
+  // This is used to optimize the special case of randomSubset where size == 2
+  override def randomTwo(precomputedCount: Int, ran1: Int, ran2: Int): Array[Agent] = {
+    // we know precomputedCount, or this method would not have been called.
+    // see randomSubset().
+    val (random1, random2) =
+      if (ran2 >= ran1)
+        // if random2 >= random1, we need to increment random2 to choose a later agent.
+        (ran1, ran2 + 1)
+      else
+        (ran2, ran1)
+    if (!kind.mortal)
+      Array(
+        array(random1),
+        array(random2))
+    else {
+      val it = iterator
+      var i = 0
+      // skip to the first random place
+      while(i < random1) {
+        it.next()
+        i += 1
+      }
+      Array(it.next(), {
+        // skip to the second random place
+        i += 1
+        while (i < random2) {
+          it.next()
+          i += 1
+        }
+        it.next()
+      })
+    }
+  }
+
+  override def randomSubsetGeneral(resultSize: Int, precomputedCount: Int, random: MersenneTwisterFast) = {
+    val result = new Array[Agent](resultSize)
+    if (precomputedCount == array.size) {
+      var i, j = 0
+      while (j < resultSize) {
+        if (random.nextInt(precomputedCount - i) < resultSize - j) {
+          result(j) = array(i)
+          j += 1
+        }
+        i += 1
+      }
+    }
+    else {
+      val iter = iterator
+      var i, j = 0
+      while (j < resultSize) {
+        val next = iter.next()
+        if (random.nextInt(precomputedCount - i) < resultSize - j) {
+          result(j) = next
+          j += 1
+        }
+        i += 1
+      }
+    }
+    result
+  }
+
+  /// iterator methods
+
+  // returns an Iterator object of the appropriate class
+  override def iterator: AgentIterator =
+    if (!kind.mortal)
+      new Iterator
+    else
+      new DeadSkippingIterator
+
+  // shuffling iterator = shufflerator! (Google hits: 0)
+  // Update: Now 5 Google hits, the first 4 of which are NetLogo related,
+  // and the last one is a person named "SHUFFLER, Ator", which Google thought
+  // was close enough!  ;-)  ~Forrest (10/3/2008)
+
+  override def shufflerator(rng: MersenneTwisterFast): AgentIterator =
+    // note it at the moment (and this should probably be fixed)
+    // Job.runExclusive() counts on this making a copy of the
+    // contents of the agentset - ST 12/15/05
+    new Shufflerator(rng)
+
+  /// iterator implementations
+
+  private class Iterator extends AgentIterator {
+    protected var index = 0
+    override def hasNext = index < array.size
+    override def next() = {
+      val result = array(index)
+      index += 1
+      result
+    }
+  }
+
+  // extended to skip dead agents
+  private class DeadSkippingIterator extends Iterator {
+    // skip initial dead agents
+    while (index < array.size && array(index).id == -1)
+      index += 1
```

Comment:
Since `hasNext` is just `index < array.size`, it expects `index` to already be pointing at the next element. Thus, pre-initialization is needed. The solution would be to make `hasNext` do the work of getting the next agent. In my experience, that's actually the standard way to writer iterators. Performance wise, I would think they'd be identical, but obviously I'd have to measure. I think I will leave it as is for now.